### PR TITLE
Feature Magento2 client

### PR DESCRIPTION
**Pull request by curzio-della-santa**

### CODE REVIEW

File: examples/magento/getProductStock.php [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+}
+
+$response = $client->getProductStock(
+    array(
+        'Authorization' => sprintf('Bearer %s', $token),
+        'productSku' => $id,
+    )
+);
+
+
```
Comment (ivan-wolf) [inline examples/magento/getProductStock.php]: Remove one empty line

File: examples/magento/getProductStock.php [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+use Denner\Client\MagentoClient;
+
+$config = require realpath(__DIR__ . '/../bootstrap.php');
+$params = array(
+);
+
+$client = MagentoClient::factory($config);
+
+$token = @$_GET['token'] ?: 'ajs8vvnm0vgk976iiqv6optdsjucp05r';
```
Comment (ivan-wolf) [inline examples/magento/getProductStock.php]: Does it make sense to have a default token here? I recommend requiring it as URL param.

File: examples/magento/getProductStock.php [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+$id = @$_GET['id'] ?: '050005';
```
Comment (ivan-wolf) [inline examples/magento/getProductStock.php]: Does it make sense to have a default article ID here? I recommend requiring it as URL param.

File: examples/magento/getProductStock.php [hunk 4]
```diff
@@ -0,0 +1,27 @@
+<?php
+
+use Denner\Client\MagentoClient;
+
+$config = require realpath(__DIR__ . '/../bootstrap.php');
+$params = array(
```
Comment (ivan-wolf) [inline examples/magento/getProductStock.php]: Never used?

File: examples/magento/getToken.php [hunk 5]
```diff
@@ -0,0 +1,43 @@
+<?php
+
+use Denner\Client\MagentoClient;
+
+$config = require realpath(__DIR__ . '/../bootstrap.php');
+$params = array(
```
Comment (ivan-wolf) [inline examples/magento/getToken.php]: Never used?

File: examples/magento/getToken.php [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+
+//$i = 0;
+//$token = '';
+//
+//// Result is a plain string, not JSON as should be.
+//// Can't get resource (results in error for the malformed body), but can get each byte of the body
+//while (($c = $response->offsetGet($i++)) !== null) {
+//   $token .= $c;
+//}
```
Comment (ivan-wolf) [inline examples/magento/getToken.php]: Remove commented code?

File: examples/magento/getToken.php [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+$token = $response->getResource()->get('response');
+
+var_dump('Token:', $token);
+
+$url = "http://" . $_SERVER['SERVER_NAME'] . dirname($_SERVER['REQUEST_URI']);
+
+$operations = array(
```
Comment (ivan-wolf) [inline examples/magento/getToken.php:22]: What's the purpose of this?

File: examples/magento/listOrders.php [hunk 8]
```diff
@@ -0,0 +1,69 @@
+<?php
+
+use Denner\Client\MagentoClient;
+
+$config = require realpath(__DIR__ . '/../bootstrap.php');
+$params = array(
```
Comment (ivan-wolf) [inline examples/magento/listOrders.php]: Never used?

File: examples/magento/listOrders.php [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+                        )
+                    )
+                ),
+            ),
+            'pageSize' => '5',
+        ),
+    )
+);
+
+
```
Comment (ivan-wolf) [inline examples/magento/listOrders.php]: Remove one empty line

File: examples/magento/listOrders.php [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+var_dump($response->getResource());
+// echo '<pre>'; print_r($response->getResource()->get('items'));
+
+$orders = array();
```
Comment (ivan-wolf) [inline examples/magento/listOrders.php:52]: Add empty line

File: examples/magento/listProducts.php [hunk 11]
```diff
@@ -0,0 +1,60 @@
+<?php
+
+use Denner\Client\MagentoClient;
+
+$config = require realpath(__DIR__ . '/../bootstrap.php');
+$params = array(
```
Comment (ivan-wolf) [inline examples/magento/listProducts.php]: Never used?

File: examples/magento/listProducts.php [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+$client = MagentoClient::factory($config);
+
+$token = @$_GET['token'] ?: 'ajs8vvnm0vgk976iiqv6optdsjucp05r';
+
+if (!$token) {
+    throw new RuntimeException('Missing or invalid parameter "token"');
+}
+
+
+var_dump($token);
```
Comment (ivan-wolf) [inline examples/magento/listProducts.php]: Remove one empty line

File: examples/magento/getProductStock.php [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+$token = @$_GET['token'] ?: 'ajs8vvnm0vgk976iiqv6optdsjucp05r';
+$id = @$_GET['id'] ?: '050005';
+
+if (!$token) {
+    throw new RuntimeException('Missing or invalid parameter "token"');
+}
+
+$response = $client->getProductStock(
+    array(
+        'Authorization' => sprintf('Bearer %s', $token),
```
Comment (ivan-wolf) [inline examples/magento/getProductStock.php:21]: Wouldn't it be better to handle authorization client-internally?

The client could automatically retrieve a token before the first call.

File: examples/magento/listOrders.php [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+    throw new RuntimeException('Missing or invalid parameter "token"');
+}
+
+$response = $client->listOrders(
+    array(
+        'Authorization' => sprintf('Bearer %s', $token),
+        'searchCriteria' => array(
+            'filterGroups' => array( // FilterGroups are connected with AND
+                array(
+                    'filters' => array( // FilterGroups are connected with OR
```
Comment (ivan-wolf) [inline examples/magento/listOrders.php]: You mean Filters?

File: examples/magento/updateProductStock.php [hunk 15]
```diff
@@ -0,0 +1,42 @@
+<?php
+
+use Denner\Client\MagentoClient;
+
+$config = require realpath(__DIR__ . '/../bootstrap.php');
+$params = array(
```
Comment (ivan-wolf) [inline examples/magento/updateProductStock.php]: Never used?

File: examples/magento/updateProductStock.php [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+);
+
+$client = MagentoClient::factory($config);
+
+$token = @$_GET['token'] ?: 'ajs8vvnm0vgk976iiqv6optdsjucp05r';
+$id = @$_GET['id'] ?: '050005';
+$pid = @$_GET['pid'] ?: '666';
+$stock = (int) (@$_GET['stock'] ?: '123');
```
Comment (ivan-wolf) [inline examples/magento/updateProductStock.php]: See above comment about defaults.

File: examples/magento/updateProductStock.php [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+//$i=0;
+//$result = '';
+//
+//// Result is a plain string, not JSON as should be.
+//// Can't get resource (results in error for the malformed body), but can get each byte of the body
+//while (($c = $response->offsetGet($i++)) !== null) {
+//    $result .= $c;
+//}
+//
+//var_dump('Result:', $result);
```
Comment (ivan-wolf) [inline examples/magento/updateProductStock.php]: Remove commented code?

File: src/Denner/Client/Response/BaseResponse.php [hunk 18]
```diff
@@ -50,7 +50,17 @@ public function toArray()
...
+//                case 'application/xml; charset=utf-8':
+//                case 'application/xml':
+//                    $data = $this->xml2array($this->getHttpResponse()->xml());
+//                    break;
+//                case 'application/json; charset=utf-8':
+//                case 'application/json':
+//                default:
+                    $data = $this->getHttpResponse()->json() ?: array();
+//                    break;
+//            }
```
Comment (ivan-wolf) [inline src/Denner/Client/Response/BaseResponse.php:63]: Why did you add this and then comment it?

File: src/Denner/Client/ServiceDescription/Magento.php [hunk 19]
```diff
@@ -0,0 +1,199 @@
+<?php
+
+use Denner\Client\Response;
+
+return array(
+    'name' => 'Magento Commerce for B2B 2.2',
+    'operations' => array(
+        'integrationAdminToken' => array(
```
Comment (ivan-wolf) [inline src/Denner/Client/ServiceDescription/Magento.php]: Rename to getToken()

**Review by ivan-wolf — CHANGES_REQUESTED** [review]
Did you add tests/magento.json on purpose?

File: examples/magento/getProductStock.php (hunk 13 above)
Comment (curzio-della-santa) [reply examples/magento/getProductStock.php:21]: It was also my first thought.
But then, recalling all the SOP problems that we have with the previous Magento, I decided against it, to let the operator handle the login/authorization independently.

But to integrate this in the client is still feasible 
(The main purpose for this PR was to test that the Magento2 client can be realized and test the current access to resources we'll need)

File: examples/magento/getToken.php (hunk 7 above)
Comment (curzio-della-santa) [reply examples/magento/getToken.php:22]: Speed for testing, prepares the other examples links to easily click-and-run them (using the Authorization token generated in this request)

**Review by curzio-della-santa — COMMENTED** [review]
Thanks for review, I'll fix the coding issues asap

File: src/Denner/Client/Response/BaseResponse.php (hunk 18 above)
Comment (curzio-della-santa) [reply src/Denner/Client/Response/BaseResponse.php:63]: Initially, I had to set the Accept header to 'application/xml' for requests, so that the response Content-Type was XML to parse the token authorization answer (that is plain text on JSON, but correct with XML).

But the short xml2array function I implemented here is not good at all, should use a professional library/deserialzer for this.

Only in a second pass I decided to introduce a new Response\PlainTextResponse to bypass the incorrect response returned by Magento 2, so that I could set the Accept header back to JSON.

But the extension to accept XML would be a correct enhancement for this library. Therefore I did only commented it out.
